src/cuCompactor.cuh

```diff
@@ -46,37 +48,39 @@ __global__ void compactK(T* d_input,int length, T* d_output,int* d_BlocksOffset,
 		int pred = predicate(d_input[idx]);
 		int w_i = threadIdx.x/warpSize; //warp index
 		int w_l = idx % warpSize;//thread index within a warp
-		int t_m = INT_MAX >> (warpSize-w_l-1); //thread mask (ERROR IN THE PAPERminus one is required)
 
-		int b	= __ballot(pred) & t_m; //balres = number whose ith bit isone if the ith's thread pred is true masked up to the current index in warp
+		// compute exclusive prefix sum based on predicate validity to get output offset for thread in warp
+		int t_m = FULL_MASK >> (warpSize-w_l); //thread mask
+		int b	= __ballot_sync(FULL_MASK,pred) & t_m; //balres = number whose ith bit is one if the ith's thread pred is true masked up to the current index in warp
 		int t_u	= __popc(b); // popc count the number of bit one. simply count the number predicated true BEFORE MY INDEX
 
-
-
+		// last thread in warp computes total valid counts for the warp
 		if(w_l==warpSize-1){
 			warpTotals[w_i]=t_u+pred;
 		}
 
-
+		// need all warps in thread block to fill in warpTotals before proceeding
 		__syncthreads();
 
-
-		if(w_i==0 && w_l<blockDim.x/warpSize){
+		// first numWarps threads in first warp compute exclusive prefix sum to get output offset for each warp in thread block
+		int numWarps = blockDim.x/warpSize;
+		unsigned int numWarpsMask = FULL_MASK >> (warpSize-numWarps);
+		if(w_i==0 && w_l<numWarps){
 			int w_i_u=0;
-			for(int j=0;j<=5;j++){
-				int b_j =__ballot( warpTotals[w_l] & pow2i(j) ); //# of the ones in the j'th digit of the warp offsets
+			for(int j=0;j<=5;j++){ // must include j=5 in loop in case any elements of warpTotals are identically equal to 32
+				int b_j =__ballot_sync(numWarpsMask, warpTotals[w_l] & pow2i(j) ); //# of the ones in the j'th digit of the warp offsets
```

Comment:
Same fix needed here for older version of CUDA

Reply:
Agreed, I will make this change.